### ISSUE TITLE
Pylance ignores __new__ return type for reportAbstractUsage

### ISSUE THREAD

**Issue by Pierre-Bartet** [microsoft/pyright#10138]
When using a factory pattern where a `Base` abstract class returns a non abstract object using `__new__`, `Pylance` wrongly reports a `reportAbstractUsage` error, because it somewhat thinks that an abstract class is being instantiated.

What is weird is that when doing `x = Base(2)`, `Pylance` report the correct (non abstract) type for `x`, but still complain about `Base(2)` being an instantiation of an abstract class.

## Environment data

-   Pylance version: 2025.3.2
-   Python version: 3.10.12

## Code Snippet

```python
from typing import Union
from abc import ABC, abstractmethod

class Base(ABC):
    def __new__(cls, c) -> Union["Derived1", "Derived2"]:
        if c == 1:
            return super(cls, Derived1).__new__(Derived1)
        
        if c == 2:
            return super(cls, Derived2).__new__(Derived2)
        
        raise ValueError
    
    @abstractmethod
    def do_something(self) -> int: ...
    

class Derived1(Base):
    def do_something(self) -> int:
        return 1
    

class Derived2(Base):
    def do_something(self) -> int:
        return 2
    

# Here Pylance cries because: Cannot instantiate abstract class "Base"
#  "Base.do_something" is not implemented
# Yet the type reported for `x` is correct here (Union["Derived1", "Derived2"])
x = Base(2)  

```

**Comment by Pierre-Bartet** [comment]
I agree that __new__ should be avoided most of the time, but it exists and it is used.

I don't see how pyright would be right here, at no point is an abstract Base object being built.

Here I put an example with derived classes, but it is the same whatever the return type of __new__.

The correct thing is just to use the return type of __new__, as it is done for inferring the type of x.

**Comment by lukas-lang** [comment]
It's really a shame that type checkers refuse to acknowledge that `__new__` is a thing in Python. I recently had to disable mypy in my project because it plays even less nicely with returning subclasses from `__new__` than Pyright does.

As @Pierre-Bartet noted, the code works as expected, and Pyright even understands that `x` will be of a non-abstract type. The error message is even produced when the type of `x` can be fully deduced (by providing appropriate overloads of `__new__`):

```python
from typing import Union, overload, Literal
from abc import ABC, abstractmethod

class Base(ABC):
    @overload
    def __new__(cls, c: Literal[1]) -> "Derived1": ...
    @overload
    def __new__(cls, c: Literal[2]) -> "Derived2": ...

    def __new__(cls, c) -> Union["Derived1", "Derived2"]:
        if c == 1:
            return super(cls, Derived1).__new__(Derived1)
        
        if c == 2:
            return super(cls, Derived2).__new__(Derived2)
        
        raise ValueError

    @abstractmethod
    def do_something(self) -> int: ...
    

class Derived1(Base):
    def do_something(self) -> int:
        return 1
    

class Derived2(Base):
    def do_something(self) -> int:
        return 2
    

# Here Pylance cries because: Cannot instantiate abstract class "Base"
#  "Base.do_something" is not implemented
# Yet the type reported for `x` is correct here (Derived2)
x = Base(2)  
reveal_type(x)
```

Naively, it seems to be that all that's missing is to use the type information that's already being deduced for the `reportAbstractUsage` warning.

On a more "productive" note: I managed to work around this limitation by introducing an intermediate class into the hierarchy:
```python
from typing import Union
from abc import ABC, abstractmethod

class Base(ABC):
    def __new__(cls, c) -> Union["Derived1", "Derived2"]:
        if c == 1:
            return super(cls, Derived1).__new__(Derived1)
        
        if c == 2:
            return super(cls, Derived2).__new__(Derived2)
        
        raise ValueError

class BaseHelper(Base):
    @abstractmethod
    def do_something(self) -> int: ...
    

class Derived1(BaseHelper):
    def do_something(self) -> int:
        return 1
    

class Derived2(BaseHelper):
    def do_something(self) -> int:
        return 2
    

# Here Pylance cries because: Cannot instantiate abstract class "Base"
#  "Base.do_something" is not implemented
# Yet the type reported for `x` is correct here (Union["Derived1", "Derived2"])
x = Base(2)  
reveal_type(x)
```
